### ISSUE TITLE
High quota_used_ratio after upgrading prior to 1.7.5-0

### ISSUE THREAD

**Issue by fsitedev** [tarantool/tarantool#2842]
Accidentally revealed that one of our tarantool services began to suffer from high "quota_used_ratio" values after upgrading prior to 1.7.5-0-g24b70de version.

Service has been working without reboot for two months and quota increased to >70% however before 1.7.5 we had almost static number about 2-3%.

For better understanding environment for this issue, here's a few key notes of this service usage:
- processing about 1K rps with mostly 'upserts' in memtx tables
- permanent tables rotation (create new tables, processing and droping old ones)
- very small snapshots size to prove having no "forgotten" tables nor data

**Comment by locker** [comment]
@fsitedev Which version was used before the upgrade?

**Comment by fsitedev** [comment]
We have upgraded from 1.7.4.180

**Comment by locker** [comment]
@fsitedev Thanks. Could you please also post here the output of `box.cfg` and `box.slab.info()`?

**Comment by fsitedev** [comment]
box.cfg
```log_nonblock: true
checkpoint_count: 5
rows_per_wal: 50000
vinyl_range_size: 1073741824
readahead: 16320
memtx_dir: /var/lib/tarantool/collector-node1
memtx_min_tuple_size: 16
vinyl_page_size: 8192
coredump: true
vinyl_run_size_ratio: 3.5
replication_timeout: 10
wal_max_size: 268435456
vinyl_memory: 134217728
log: /logs/tarantool/collector-node1.log
background: true
vinyl_dir: /var/lib/tarantool/collector-node1
vinyl_cache: 134217728
vinyl_read_threads: 1
vinyl_timeout: 60
listen: '9900'
checkpoint_interval: 7200
vinyl_bloom_fpr: 0.05
vinyl_max_tuple_size: 1048576
vinyl_write_threads: 2
force_recovery: false
memtx_max_tuple_size: 1048576
slab_alloc_factor: 1.05
too_long_threshold: 2
log_level: 5
wal_mode: write
worker_pool_threads: 4
hot_standby: false
read_only: false
pid_file: /var/run/tarantool/collector-node1.pid
memtx_memory: 1073741824
wal_dir: /var/lib/tarantool/collector-node1
username: tarantool
vinyl_run_count_per_level: 2
wal_dir_rescan_delay: 2
```

**Comment by fsitedev** [comment]
box.slab.info()
```
items_size: 854098104
items_used_ratio: 88.20%
quota_size: 1073741824
quota_used_ratio: 82.81%
arena_used_ratio: 87.9%
items_used: 753277448
quota_used: 889192448
arena_size: 858292200
arena_used: 754620936
```

**Comment by locker** [comment]
Could you please run the following script on your tarantool instance along with `box.slab.info()`? It just prints len/size of all spaces, should be very cheap. It will help us understand where memory goes.
```
for _, s in pairs(box.space) do has_len, len = pcall(s.len, s) print(s.id, s.name, s:bsize(), has_len and len or 0) end
```
Also, have you tried to run `collectgarbage('collect')` in Lua? If yes, does it reduce memory usage? If no, is it acceptable for you to try it in your environment?

**Comment by fsitedev** [comment]
Yes, we constantly run `collectgarbage('collect')` in our fibers and it doesn't reduce memory.
Unfortunatelly we run out of memory last friday so we had to reboot all instances. Obviously `slab.info` also has changed since the ticket-start message but dynamic persists event after reboot: we have stable `quota used` growth by few percents daily.

**Comment by fsitedev** [comment]
Here is our a little bit extended script:
``` lua
(function ()
local res = {}
  for _, s in pairs(box.space) do
    local has_len, len = pcall(s.len, s)
    table.insert(res, {id = s.id, name = s.name, bsize = s:bsize(), len = has_len and len or 0})
  end
return {slab = box.slab.info(), spaces = res}
end)()
```

**Comment by fsitedev** [comment]
Result (spaces are repeated in the output. is that normal?):
```
---
slab:
  items_size: 34900192
  items_used_ratio: 73.97%
  quota_size: 1073741824
  quota_used_ratio: 6.25%
  arena_used_ratio: 73.4%
  items_used: 25814200
  quota_used: 67108864
  arena_size: 36997240
  arena_used: 27141304
spaces:
- len: 262
  bsize: 24147
  name: socdem_1508762760
  id: 256545
- len: 262
  bsize: 24147
  name: socdem_1508762760
  id: 256545
- len: 0
  bsize: 0
  name: _vindex
  id: 289
- len: 35
  bsize: 1497
  name: _index
  id: 288
- len: 0
  bsize: 0
  name: _vpriv
  id: 313
- len: 65
  bsize: 1295
  name: site_config
  id: 195678
- len: 0
  bsize: 0
  name: _vfunc
  id: 297
- len: 1
  bsize: 29
  name: _func
  id: 296
- len: 16
  bsize: 1622
  name: _space
  id: 280
- len: 4
  bsize: 86
  name: _schema
  id: 272
- len: 4
  bsize: 103
  name: _user
  id: 304
- len: 13
  bsize: 65
  name: _truncate
  id: 330
- len: 1
  bsize: 29
  name: _func
  id: 296
- len: 13
  bsize: 174
  name: _priv
  id: 312
- len: 35
  bsize: 1497
  name: _index
  id: 288
- len: 4
  bsize: 103
  name: _user
  id: 304
- len: 1
  bsize: 40
  name: _cluster
  id: 320
- len: 1
  bsize: 40
  name: _cluster
  id: 320
- len: 13
  bsize: 174
  name: _priv
  id: 312
- len: 13
  bsize: 65
  name: _truncate
  id: 330
- len: 0
  bsize: 0
  name: _vindex
  id: 289
- len: 0
  bsize: 0
  name: _vspace
  id: 281
- len: 0
  bsize: 0
  name: _vuser
  id: 305
- len: 16
  bsize: 1622
  name: _space
  id: 280
- len: 0
  bsize: 0
  name: _vspace
  id: 281
- len: 4140
  bsize: 755459
  name: fact_1508762760
  id: 256546
- len: 4140
  bsize: 755459
  name: fact_1508762760
  id: 256546
- len: 0
  bsize: 0
  name: _vuser
  id: 305
- len: 4
  bsize: 86
  name: _schema
  id: 272
- len: 0
  bsize: 0
  name: _vpriv
  id: 313
- len: 0
  bsize: 0
  name: _vfunc
  id: 297
- len: 65
  bsize: 1295
  name: site_config
  id: 195678
...
```

**Comment by locker** [comment]
> spaces are repeated in the output. is that normal?

Yes. Each space is referenced twice in `box.space` table: by id and by name.

> Result

Thanks. So according to the data you posted, you have about 4.5K records in all your spaces which consume about 800KB while `box.slab.info()` reports that there are about 26MB of memory occupied by tuples (`items_used`). Since you claim that garbage collection doesn't help, it looks like tuples are leaking.

I haven't managed to reproduce the issue locally so far. By looking at the difference between commits 6cfefcd and 24b70de I haven't found anything suspicious, but the difference is so big that I could've easily overlooked something.

Could you please provide us with more info about your use case. I'm interested in:

- Index configuration, i.e. the output of `box.space.<space_name>` for each of your spaces.
- An example of an upsert operation you use.
- Do you use `space.truncate()`?

**Comment by fsitedev** [comment]
Our common space structure:
```
unix/:/var/run/tarantool/collector-node2.control> box.space.fact_1508838420
---
- index:
    0: &0
      unique: true
      parts:
      - type: unsigned
        fieldno: 1
      - type: string
        fieldno: 2
      - type: string
        fieldno: 3
      - type: unsigned
        fieldno: 4
      id: 0
      space_id: 514
      name: primary
      type: TREE
    primary: *0
  on_replace: 'function: 0x40e146d8'
  temporary: false
  id: 514
  engine: memtx
  enabled: true
  name: fact_1508838420
  field_count: 0
...
```

**Comment by fsitedev** [comment]
We have one cron-based lua script running once per minute and making following routines on such common space:
1. connect to tnt node and run `box.snapshot`
2. parse resulting xlogs with `xlog` package
3. drop old spaces

**Comment by fsitedev** [comment]
The most load on this space come from nginx upstream module that constantly calls one lua procedure.
Literally this procedure makes only one upsert operation. How it looks like in general: 

- Blank tuple made once on node startup with all fields
`local blank_tuple = box.tuple.new(...)`

- In procedure we prepare update and insert tuples 
`local new_update = {{'+', 10, 1}}`
`
local new_insert = {
  {'=', 1, x},
  {'=', 2, y},
  {'=', 3, z},
  {'=', 10, 1},
}
`
- Upsert in procedure at last
`space_pointer:upsert(blank_tuple:update(new_insert), new_update)`

Background fiber tracks each minute, creates new space if current minute changed and update `space_pointer` variable with newly created space

**Comment by fsitedev** [comment]
We are ready to provide more data needed to investigate the problem

**Comment by locker** [comment]
Thanks a lot. I think I managed to reproduce the problem. Looks like we may leak tuples if DML races with `box.snapshot()`. The bug seems to have been introduced between 1.7.4 and 1.7.5 with `tarantool/small` update. We are working on the fix.

**Comment by locker** [comment]
The commit that introduced the leak: https://github.com/tarantool/small/commit/b68e9323e0860877c940e090b698515d65d43dbe

**Comment by locker** [comment]
Fixed in 1.7.5-254-g685416d9
